webapp/components/AccountSettings.tsx

```diff
@@ -1,10 +1,178 @@
-import { useAppSelector } from '../hooks';
+import { getDownloadURL, ref, uploadBytes } from 'firebase/storage';
+import { useState } from 'react';
+import { ErrorDTO } from '../API/types/errors-types';
+import {
+	isUpdateUserByIdDTO,
+	isUserDTO,
+	UpdateUserByIdDTO,
+} from '../API/types/users-types';
+import { updateUser } from '../API/userHooks';
+import { useAppDispatch, useAppSelector } from '../hooks';
+import { populateUser } from '../slices/UserSlice';
 import { RootState } from '../store';
+import styles from '../styles/AccountSettings.module.scss';
+import { storage } from './Firebase/FirebaseInit';
 
 const AccountSettings = () => {
+	const dispatch = useAppDispatch();
+	const user = useAppSelector((state: RootState) => state.user?.value);
 	const userAuth = useAppSelector((state: RootState) => state.userAuth.value);
+	const [profileImage, setProfileImage] = useState<File | null>(null);
+	const [prevewImageUrl, setPrevewImageUrl] = useState<string | null>(null);
+	const [userName, setUserName] = useState<string>(user?.userName as string);
+	const [isSaveButtonDisabled, setIsSaveButtonDisabled] =
+		useState<boolean>(true);
+	const [selectedFile, setSelectedFile] = useState<File | null>(null);
 
-	return <div>{userAuth?.displayName}</div>;
+	const handleImageChange = (file: File) => {
+		if (!file) return;
+
+		setProfileImage(file);
+		setIsSaveButtonDisabled(false);
+
+		const reader = new FileReader();
+		reader.onloadend = () => {
+			setPrevewImageUrl(reader.result as string);
+		};
+		reader.readAsDataURL(file!);
+	};
+
+	const handleSave = async () => {
+		if (isSaveButtonDisabled) return;
+		if (!profileImage && userName === user?.userName) return; // return if no change
+
+		const updateUserData = {
+			userId: user?.userId as number,
+			userName: userName,
+			profileImgUrl: user?.profileImgUrl as string,
+		};
+
+		if (profileImage) {
+			// update image if a file exist, update profileImgUrl
+			try {
+				const imageFirebaseRef = `/userProfilePics/${userAuth?.email}`;
+				const snapshot = await uploadBytes(
+					ref(storage, imageFirebaseRef),
+					profileImage
+				);
+
+				const downloadURL = await getDownloadURL(snapshot.ref);
+
+				if (!downloadURL) {
+					alert('Failed to save image to firebase');
+				} else {
+					updateUserData.profileImgUrl = downloadURL;
+				}
+			} catch (error) {
+				alert('Upload unsuccessfully');
+				return;
+			}
+		}
+
+		// API call to backend
+		const updateUserRes: UpdateUserByIdDTO | ErrorDTO = await updateUser(
```

Comment:
this will likely not work! Javascript is single threaded application. When you upload the image to firebase, it won't block all other actions to wait for the image URL back. It'll execute the `updateUser` API call right away. At least I think so.

Reply:
I fixed it